### ISSUE TITLE
[RFC] Support for 2D Sprite Rendering and Animation

### ISSUE THREAD

**Issue by azriel91** [amethyst/amethyst#634]
I'd like to add support in Amethyst for 2D sprites, and to make sure I "build the right thing", I shall ask some questions in this issue. Whether I "build the thing right" can be left to the PR :sweat_smile:.

## Use Cases

Here I list some use cases so that during implementation, I don't write it in a way that disallows these to be done (or are difficult to do).

### Object - Sprite Representation

* Object is represented by an individual sprite &mdash; single entity with a sprite component.

    Example: Spaceship rendered with a single sprite.

* Object is represented by multiple sprites &mdash; multiple entities, each with a sprite component.

    Example: A character is composed from separate "parts" &mdash; body, head gear, torso, arms, legs.

### Animation

* An independent object (`B`) may be attached to another independent object (`A`) and `B`'s animation may be affected by `A`'s animation.

    Example: A weapon is held by a character.

* Multiple objects attached to a main object.

    Example: Multiple armor parts connected to the character.

* Material / Texture value mutations

    Example: Poison/power up effect changes sprite hue, invisibility gradually decreases texture alpha.

## Implementation

Here we should pick what to support (or at least, not design ourselves into a corner that it can't be added easily). Checkbox means "discussed and we know what to include / defer / exclude"

* Individual sprite image

    - [x] Should we have a separate (set of?) type(s) for a single image sprite?
    
        We could represent this as a sprite sheet with 1 row and 1 column.

        No &mdash; it's usually only used in prototyping, and a finished game almost certainly will have sheets. See [comment](https://github.com/amethyst/amethyst/issues/634#issuecomment-380568448).

* Sprite sheets

    - [x] Sprite numbering within an image.

        So far I know of these two ways, are there any more?

        + Sprites are numbered from left to right, top to bottom
        + Sprites are numbered from top to bottom, left to right

    - [x] Border between sprites.
    
        Sprites may have a border between them. In the example (#624) I just assumed it's either there or not. Does anyone think we should support:

        + a user specified border size (e.g. 5 pixels between sprites)?
        + leading border (e.g. top and left of sprite sheet starts with a border)?

        I don't know if those use cases will come up. Perhaps defer it to later a.k.a. build it if someone asks.

    - [x] Sprite numbering across images.

        Say there are 2 images, each with 5 sprites. Should we have:

        + One `Sprites` container with mesh handles `0..9`
        + Two `SpriteSheet`s with mesh handles `0..4`, and a `Sprites` with `SpriteSheetHandle`s `0..1` (intermission, does Rust do inclusive or exclusive ranges?)

        This would affect people's configuration format, e.g. for 2 sprite sheets with 2 images each:

        ```toml
        [[animations]]
          id = "Walk"
          next = "Stand"
          frames = [
            { sheet = 0, sprite = 0, wait = 2 },
            { sheet = 0, sprite = 1, wait = 2 },
            { sheet = 1, sprite = 0, wait = 2 },
            { sheet = 1, sprite = 1, wait = 2 },
          ]
        ```

        vs

        ```toml
        [[animations]]
          id = "Walk"
          next = "Stand"
          frames = [
            { sprite = 0, wait = 2 },
            { sprite = 1, wait = 2 },
            { sprite = 2, wait = 2 },
            { sprite = 3, wait = 2 },
          ]

        ```

        _(don't cry, the configuration markup is application specific)_

        _That configuration format is just me spiking something for my game, whose animations may be defined using ticks (frames) instead of keyframe times, but that's my game's concern to translate into Amethyst animation_

    The above 3 points can be all addressed by making this user defined. See [comment](https://github.com/amethyst/amethyst/issues/634#issuecomment-380568448).

* Animation

    - Picking which sprites to animate

        + User specifies sprite number for each frame.

            This one is probably the most important, as the next two variants can be accomplished with this implementation, whereas you're not guaranteed to be able to do this one with the next two.

            This also allows the same sprite to be used in different animations.

        + User specifies start and end frame numbers.
        + Entire sheet (i.e. the whole sheet represents one animation)

        This one is probably going to be user defined, depending on the sprite sheet types. Shall code and see what happens here since I'm not that familiar with the codebase yet.

**Comment by Rhuagh** [comment]
I would say that spritesheets shouldn't really be represented by multiple meshes, you can easily have single quad mesh for all sprites, and just use indexing into a texture, using texture coordinates.

For animation, we need to define what types of animation we'd like to support. The first, and easiest is just do be able to define what sprite in a sheet to show at indiviual keyframes, so in this the animation would be a step animation controlling a texture index basically (or possible texture coordinates).

**Comment by Xaeroxe** [comment]
> Should we have a separate (set of?) type(s) for a single image sprite?

I'd say probably not.  It's a common need for prototyping, but for a finished game I think you'll almost always want to convert this to a full animation, and we should make such a conversion as easy as possible.

> Sprite numbering within an image.

I personally would actually go at this in a completely different way.  Sprite positioning, size, and ordering should be 100% user defined.  This is more complicated to use but it has a few benefits:

 -  Easy frame re-use
 - All animations for a character can be stored in a single sheet, which improves memory allocation
 - Sloppy artists are easy to workaround.  Borders? who cares?  Sprite is awkwardly offset by a pixel or two for a few frames?  Easy to fix without opening up picture editing software.

Of course this has the downside of making importers slightly more complicated, but it comes with enough benefits that I feel it's worthwhile.

**Comment by azriel91** [comment]
Closed by #638
